File: labo1/exercise1_et_2/bin.py
def cree_nombre_bin(number):
    lst = []
    rest = number
    while rest > 0:
        lst.append(rest % 2)
        rest = rest >> 1
    if lst == []:
        lst.append(0)

    return lst

def remplir_zero(lst, supplement):
    rest = supplement - len(lst)
    for i in range (rest):
        lst.append(0)


def initialiser_list_vide(taille):
    lst = []
    for i in range(taille):
        lst.append(0)
    return lst

def add_bin(A,B):
    C = initialiser_list_vide(len(A) + 1)
    retenue = 0
    n = len(B)
    for i in range(n):
        C[i] = (A[i] + B[i] + retenue) % 2
        retenue = (A[i] + B[i] + retenue) >> 1
    C[n] = retenue
    return C

def mult_bin(A,B):
    C = initialiser_list_vide(len(A))
    decalage = 0
    n = len(B)
    for i in range(n):
        temp = initialiser_list_vide(n + decalage)
        for j in range(n):
            temp[j + decalage] = A[j] * B[decalage]
        C = add_bin(temp, C)
        decalage = decalage + 1
    return C


def afficher_nombre_binaire(lst):
    aff = lst
    aff.reverse()
    print(*aff,sep="",end=" ")


nombre1 = int(input("Entrer un premier nombre à additionner : "))
nombre2 = int(input("Entrer un deuxième nombre à additionner : "))
op = input("Opération à entrer entre + et * : ")

if op != "+" and op != "*":
    print("Opérateur inconnu")
else:


    lst_bin1 = cree_nombre_bin(nombre1)
    lst_bin2 = cree_nombre_bin(nombre2)
    if len(lst_bin1) > len(lst_bin2):
        remplir_zero(lst_bin2,len(lst_bin1))
    else:
        remplir_zero(lst_bin1,len(lst_bin2))

    afficher_nombre_binaire(lst_bin1)
    print(op,end=" ")
    afficher_nombre_binaire(lst_bin2)
    print("=",end=" ")
    lst_bin1.reverse()
    lst_bin2.reverse()

    if op == "+":
        resultat = add_bin(lst_bin1,lst_bin2)
    else:
        resultat = mult_bin(lst_bin1,lst_bin2)
    afficher_nombre_binaire(resultat)
    print()

File: labo1/exercise1_et_2/bin.cpp
#include <iostream>
#include <string>
#include <vector>

void afficherNombreBinaire(const std::vector<unsigned int> num) {
    for (unsigned int i = num.size(); i > 0; i--) {
        std::cout << num.at(i - 1);
    }
}

void remplirZero(std::vector<unsigned int>& num, unsigned int n) {
    unsigned int rest = n - num.size();
    for (unsigned int i = 0; i < rest; i++) {
        num.push_back(0);
    }
}

std::vector<unsigned int> creeNombreBin(const unsigned int num) {
    unsigned int rest = num;
    std::vector<unsigned int> resultat;
    if (num == 0)
        resultat.push_back(0);
    else {
        while (rest > 0) {
            resultat.push_back(rest % 2);
            rest >>= 1;
        }
    }

    return resultat;
}

std::vector<unsigned int> initZero(unsigned int n) {
    std::vector<unsigned int> tab(n);
    for (unsigned int i = 0; i < n; i++) {
        tab[i] = 0;
    }
    return tab;
}

std::vector<unsigned int> addBin(std::vector<unsigned int> A, std::vector<unsigned int> B) {
    std::vector<unsigned int> C = initZero(A.size() + 1);
    unsigned int retenue = 0;
    for (unsigned int i = 0; i < A.size(); i++) {
        C[i] = (A[i] + B[i] + retenue) % 2;
        retenue = (A[i] + B[i] + retenue) >> 1;
    }
    C[A.size()] = retenue;
    return C;
}

std::vector<unsigned int> multBin(std::vector<unsigned int> A, std::vector<unsigned int> B) {
    std::vector<unsigned int> resultat = initZero(A.size());
    unsigned int decalage = 0;
    for (unsigned int i = 0; i < B.size(); i++) {
        std::vector<unsigned int> temp = initZero(A.size() + decalage);
        for (unsigned int j = 0; j < A.size(); j++) {
            temp[j + decalage] = A[j] * B[decalage];
        }
        resultat = addBin(temp, resultat);
        decalage++;
    }
    return resultat;
}

int main(int argc, char** argv) {
    if (argc != 4) {
        std::cerr << "bin num num <add|mul>\n";
        return -1;
    }
    unsigned int num1 = atoi(argv[1]);
    unsigned int num2 = atoi(argv[2]);
    std::string op = argv[3];

    std::vector<unsigned int> A = creeNombreBin(num1);
    std::vector<unsigned int> B = creeNombreBin(num2);
    std::vector<unsigned int> C;

    if (A.size() > B.size())
        remplirZero(std::ref(B), A.size());
    else
        remplirZero(std::ref(A), B.size());

    if (op == "add")
        C = addBin(A, B);

    else if (op == "mul")
        C = multBin(A, B);
    else {
        std::cerr << "opération non reconnu\n";
        return -1;
    }

    std::cout << "(";
    afficherNombreBinaire(A);
    if (op == "add")
        std::cout << ")_2 + (";
    else
        std::cout << ")_2 * (";
    afficherNombreBinaire(B);
    std::cout << ")_2 = (";
    afficherNombreBinaire(C);
    std::cout << ")_2\n";

    return 0;
}
